### PR TITLE
Leveling

### PR DESCRIPTION
**Pull request by ToxicMushroom**
replace #28

### CODE REVIEW

File: bot/src/main/kotlin/me/melijn/bot/commands/LevelingExtension.kt [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+        "missing"
+    } else {
+        val entryUser = shardManager.retrieveUserById(userId).awaitOrNull()
+        if (entryUser == null) {
+            missingUserManager.markUserDeleted(userId)
+            "missing"
+        } else {
+            entryUser.effectiveName
+        }
+    }
```
Comment (zeroeightysix) [inline bot/src/main/kotlin/me/melijn/bot/commands/LevelingExtension.kt:418]: code duplication

File: bot/src/main/kotlin/me/melijn/bot/commands/LevelingExtension.kt [hunk 2]
```diff
@@ -0,0 +1,605 @@
...
+
+    enum class LeaderboardOpt : InferredChoiceEnum {
+        GuildXP, GlobalXP, Currency
+    }
+
+    inner class LeaderboardArgs : Arguments() {
+        val options by enumChoice<LeaderboardOpt> {
+            name = "option"
+            description = "5"
+            typeName = "5"
```
Comment (zeroeightysix) [inline bot/src/main/kotlin/me/melijn/bot/commands/LevelingExtension.kt:519]: 5

File: bot/src/main/kotlin/me/melijn/bot/database/manager/BalanceManager.kt [hunk 3]
```diff
@@ -39,4 +43,63 @@ class BalanceManager(driverManager: DriverManager) : AbstractUserBalanceManager(
         }
         driverManager.removeCacheEntry("$classKeyPrefix:${flake.id}") // drop cache due to possible race conditions
     }
-}
\ No newline at end of file
+
+
+
+    suspend fun getTop(count: Int, offset: Long): List<AugmentedBalanceData> {
```
Comment (zeroeightysix) [inline bot/src/main/kotlin/me/melijn/bot/database/manager/BalanceManager.kt:49]: newlines

File: bot/src/main/kotlin/me/melijn/bot/commands/LevelingExtension.kt [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+                val xp = xpManager.getGlobalXP(target)
+                val guildXp = guild?.let { xpManager.getGuildXP(it, target) }
+                respond {
+                    val bufferedImage = LevelingExtension::class.java.getResourceAsStream("/slice2.png").use {
+                        ImmutableImage.wrapAwt(ImageIO.read(it))
+                    }.awt()
+                    val bars = drawXpCard(bufferedImage, xp, guildXp, target)
+
+                    val baos = ByteArrayOutputStream()
+                    ImageIO.write(bars, "png", baos)
```
[the remark this answers is not in the document]
Comment (ToxicMushroom) [reply bot/src/main/kotlin/me/melijn/bot/commands/LevelingExtension.kt:72]: why should I expect failure here ?

**Comment by ToxicMushroom** [comment]
Fixed on of the #26 entries

**Pull request merged**